Fixing issue where generic model types wouldn't load properly with the pipeline

Narsil opened the pull request:
# What does this PR do?

When this occurs https://github.com/huggingface/transformers/issues/17929
we can provide a better error message since this is detectable at load time
and the fix should happen within `transformers`.

Found out 3 odd cases which have been dealt with differently:

- `translation` actually uses `translation_XX_to_YY` and also relies on `task_specific_params` for some model configs.
  I tried cleaning that up and using `task_specific_params` only once, but the rabbit hole is deep, and it would have meant more
  code changes that this PR should hold. Waiting for a subsequent PR.
  The issue is that `translation_XX_to_YY` is not a normalized task name and is not within `NO_TOKENIZER_TASKS` nor `NO_FEATURE_EXTRACTION_TASKS` so the configuration on wether we should load or not doesn't work.
- `feature-extraction`. That one is extremely special, since ALL models could in theory use that pipeline, and so we cannot enforce or detect anything statically on what should be loaded or not.
- `automatic-speech-recognition` has this `speech-encoder-decoder` type of model, which do not define any `tokenizer` class, so the `type(config)` is NOT within `TOKENIZER_MAPPING` (correctly), but the first version of the check would fail when deciding staticly if we should load the tokenizer or not. The fix was to check if the user passed a tokenizer or not (if tokenizer is passed we should never try to do anything anyway)

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@sugger

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

Narsil committed:
Adding a better error message when the model is improperly configured

within transformers.

Narsil commented on src/transformers/pipelines/__init__.py:
```diff
@@ -667,6 +667,15 @@ def pipeline(
+
+    if not load_feature_extractor and task not in NO_FEATURE_EXTRACTOR_TASKS:
+        raise EnvironmentError(
+            f"There is a problem in `transformers`. The task {task} requires a feature extractor, however the model {type(model_config)} seems to not support feature-extractors. Please report this issue. This is likely a misconfiguration in the library, please report this issue."
```
```suggestion
            f"There is a problem in `transformers`. The task {task} requires a feature extractor, however the model {type(model_config)} seems to not support feature-extractors. This is likely a misconfiguration in the library, please report this issue."
```

Narsil committed:
Update src/transformers/pipelines/__init__.py

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18392). All of your documentation changes will be reflected on that endpoint.

Narsil committed:
Black version.

Narsil committed:
Overriding task aliases so that tokenizer+feature_extractor

values are correct.

Narsil committed:
Fixing task aliases by overriding their names early

Narsil committed:
X.

Narsil committed:
Fixing feature-extraction.

Narsil committed:
black again.

Narsil committed:
Normalizing `translation` too.

Narsil committed:
Fixing last few corner cases.

translation need to use its non normalized name (translation_XX_to_YY,
so that the task_specific_params are correctly overloaded).
This can be removed and cleaned up in a later PR.

`speech-encode-decoder` actually REQUIRES to pass a `tokenizer` manually
so the error needs to be discarded when the `tokenizer` is already
there.

Narsil committed:
doc-builder fix.

sgugger commented on src/transformers/pipelines/__init__.py:
```diff
@@ -661,12 +662,37 @@ def pipeline(
+        and task != "feature-extraction"
+    ):
+        raise EnvironmentError(
+            f"There is a problem in `transformers`. The task {task} requires a feature extractor, however the model"
+            f" {type(model_config)} seems to not support feature-extractors. This is likely a misconfiguration in the"
+            " library, please report this issue."
```
Same here.

sgugger commented on tests/pipelines/test_pipelines_zero_shot_image_classification.py:
```diff
@@ -172,6 +172,22 @@ def test_small_model_tf(self):
             ],
         )
 
+    @slow
+    @require_torch
+    def test_large_model_misconfigured(self):
+        # XXX this should be a fast test, but the triggering arch
+        # VisionTextDualEncoderModel is missing for small tests
+        # https://huggingface.co/hf-internal-testing
+        # This test will also start to fail, once this architecture
+        # correctly defines AutoFeatureExtractor. At this point
+        # we can safely remove this test as we don't really want
+        # to keep around an invalid model around just for this.
+        with self.assertRaises(EnvironmentError):
+            pipeline(
+                task="zero-shot-image-classification",
+                model="Bingsu/vitB32_bert_ko_small_clip",
+            )
```
I wouldn't include this test as it is. It's easy to build a small model in hf-internal-testing that is misconfigured and does not provide a tokenizer class while using a generic encoder/decoder arch.

Narsil replied:
True.

sgugger reviewed (commented):
Thanks for your PR! I'm all for better and clearer error messages but I want to emphasize this is by no way a failure of the library. All the Encoder/Decoder classes are generic and can be uses with multiples different encoders/decoders. Therefore, they do not have a default tokenizer/feature extractor class associated to them and it's up to the users to set those in the `config.json` (or `tokenizer_config.json`/`preprocessor_config.json`) so the Auto classes then work properly.

Thus added some tweaks.

Narsil commented on src/transformers/pipelines/__init__.py:
```diff
@@ -661,12 +662,37 @@ def pipeline(
+        and normalized_task not in NO_TOKENIZER_TASKS
+        and task != "feature-extraction"
+    ):
+        raise EnvironmentError(
+            f"There is a problem in `transformers`. The task {task} requires a tokenizer, however the model"
+            f" {type(model_config)} seems to not support tokenizer. This is likely a misconfiguration in the library,"
+            " please report this issue."
```
Okay ! I knew about the speech ones, not the vision ones. So it's the same.

Narsil committed:
Fixing the real issue.

Narsil committed:
Removing dead code.

Narsil commented:
It seems I had completely misunderstood what was going on there. I thought it was misconfiguration while it's more of a normal state of things (Wasn't aware we had added those for vision too).

My new proposed PR then actually fixes the underlying issue initially created #17929 . 

The way I did it is keep some manual bookkeeping for these "multi model" configurations (is the name right) ?

Then if we are actually using one of these models, attempt to load the `Tokenizer`/`feature_Extractor` looking exclusively at whether or not the task requires one.
This should fix the original issue, and it so happens we had a test that could easily be updated to support those use cases.

What do you think of this approach ? 

If a user created a model and forgot to upload either one of the necessary components, the pipeline will simply fail to load attempting to load one of them. I think that sort of failure mode should be OK to understand and users should be able to recover on their own. So no need for error messages now.

I am still keeping the regular way to detect if we need the tokenizer for other types of configs, but then we will still fail if the AutoTokenizer/FeatureExtractor is not correctly configured. I think maybe switching entirely to `NO_TOKENIZER_TASKS` detection seems easier in the long run but I didn't want to do such a change in a small PR. (`feature-extraction` will still be a corner case)

sgugger commented on src/transformers/pipelines/__init__.py:
```diff
@@ -25,6 +25,8 @@
 
 from numpy import isin
 
+from transformers import SpeechEncoderDecoderConfig, VisionTextDualEncoderConfig
```
Let's not add new dependencies between pipeline and specific models if we can avoid it. Here we can detect the proper model types I believe.

If those are kept, they should be relative imports like the rest.

Narsil replied:
Perfectly understandable. It's done.

sgugger commented on src/transformers/pipelines/__init__.py:
```diff
@@ -661,12 +665,34 @@ def pipeline(
+        tokenizer is None
+        and not load_tokenizer
+        and normalized_task not in NO_TOKENIZER_TASKS
+        and type(model_config) in MULTI_MODEL_CONFIGS
```
Here we could use `model_config.model_type` to avoid importing the config classes, or use `model_config.__class__.___name__` to compare with strings instead of objects.

sgugger commented on src/transformers/pipelines/__init__.py:
```diff
@@ -661,12 +665,34 @@ def pipeline(
+        feature_extractor is None
+        and not load_feature_extractor
+        and normalized_task not in NO_FEATURE_EXTRACTOR_TASKS
+        and type(model_config) in MULTI_MODEL_CONFIGS
```
Same here.

sgugger reviewed (commented):
Thanks for iterating on your PR. LGTM with just the problem of importing from specific models in the pipeline file. I'd prefer to avoid it so the library stay as compartmentalized as possible. Made some suggestions of alternatives.

Narsil committed:
Do not import the actual config classes.
